tests/net-message_test.cpp

```diff
@@ -0,0 +1,353 @@
+}
+void PrintTo(const Message& msg, std::ostream* os) {
+	auto old_flags = os->flags();
+	*os << "Message{len=" << msg.GetBodyLength() << ", type=" << msg.GetMessageType() << ", id=" << std::hex
+		<< msg.GetMessageID() << std::dec << ", seq=" << msg.GetSequence() << "}";
+	os->setf(old_flags);
+}
+// make something sufficiently large to require multiple messages
+static std::string GetLongTestString(size_t num_messages, bool variant = false) {
+	using namespace std::string_literals;
+	std::string testdata{"TEST0123456789abcdefghijklmnopqrstuvwxyz"s};
+	if (variant) {
+		testdata = "TEST0123456789ZYXWVUTSRQPONMLKJIHGFEDCBA"s;
+	}
+	std::string teststring;
+	teststring.append(testdata);
+	while (teststring.size() < (Message::max_body_length * num_messages)) {
+		teststring.append(testdata);
+	}
+	return std::move(teststring);
+}
+
+static void MessageListContentEqual(
+		MessagePool::list_type& list_a, MessagePool::list_type& list_b, bool pointers_should_equal = false) {
+	EXPECT_NE(&list_a, &list_b);
+	auto iterator_a = list_a.begin();
+	auto iterator_b = list_b.begin();
+	while (iterator_a != list_a.end() && iterator_b != list_b.end()) {
+		auto& item_a = *iterator_a;
+		auto& item_b = *iterator_b;
+		if (pointers_should_equal) {
+			EXPECT_EQ(&*item_a, &*item_b);
+		}
+		else {
+			EXPECT_NE(&*item_a, &*item_b);
+		}
+		EXPECT_PRED2(HeadersEqual, *item_a, *item_b);
+		std::string value_a(item_a->GetBodyPTR(), item_a->GetBodyLength());
+		std::string value_b(item_b->GetBodyPTR(), item_b->GetBodyLength());
+		EXPECT_EQ(value_a, value_b);
+		iterator_a++;
+		iterator_b++;
+	}
+	bool lists_have_same_number_items = iterator_a == list_a.end() && iterator_b == list_b.end();
+	EXPECT_TRUE(lists_have_same_number_items);
+}
+
+TEST(Message, limits) {
+	Message msg;
+	auto teststring = GetLongTestString(2);
+	// ensure that our test string requires more than one message (it should)
+	EXPECT_GT(teststring.size(), Message::max_body_length);
+	// does it start empty?
+	EXPECT_EQ(msg.GetBodyLength(), 0);
+	msg.SetBodyLength(Message::max_body_length);
+	EXPECT_EQ(msg.GetBodyLength(), Message::max_body_length);
+	// can it get too big?
+	msg.SetBodyLength(Message::max_body_length + 1);
+	EXPECT_EQ(msg.GetBodyLength(), Message::max_body_length);
+	msg.SetBodyLength(Message::max_body_length * 2);
+	EXPECT_EQ(msg.GetBodyLength(), Message::max_body_length);
+	// can we shrink?
+	msg.SetBodyLength(0);
+	EXPECT_EQ(msg.GetBodyLength(), 0);
+}
+TEST(Message, reflexive) {
+	Message msg_a;
+	Message msg_b;
+	msg_a.SetMessageType(MessageType::ENTITY_CREATE);
+	msg_a.SetBodyLength(4);
+	std::string test{"TEST"};
+	memcpy(msg_a.GetBodyPTR(), test.data(), test.size());
+	msg_a.encode_header();
+	auto buf = asio::buffer(msg_a.GetDataPTR(), msg_a.length());
+	memcpy(msg_b.GetDataPTR(), buf.data(), buf.size());
+	msg_b.decode_header();
+	EXPECT_EQ(msg_b.GetBodyLength(), msg_a.GetBodyLength());
```

Comment:
bit of a note to self, I could use the headers compare predicate here instead, and probably fuzz these better